### PR TITLE
PR Chapter 02 - Introducing Services and the ServiceManager

### PR DESCRIPTION
**Pull request by manuakasam**
Discussion about the second chapter goes here :)

### CODE REVIEW

**Comment by tfountain** [comment]
General point on the use of interfaces here: 

In my own (userland) code, I tend only to use interfaces in code that may potentially be used or extended by other people, and in situations where I can at least vaguely see use cases for alternate implementations of said class. This chapter implies that you should create interfaces for all your service classes and all your models.

Whilst it's great that we're pushing best practices in this guide, I'm not sure if this necessarily is a best practice, and if it is, I think we need to justify the use of interfaces a little better. (This may be covered in a subsequent chapter - I've not read ahead.)

If I was creating a module like this myself, I doubt I would create either of these interfaces. Perhaps if after writing the module I intended to open source the code I might put in an interface for the service class to make it easier for someone to pull data in from alternate data sources, but even then I doubt I'd create one for the model, as I can't think of a situation where an interface would be better here than simply extending the class.

I'm all for best practices, but we need to justify them if their purpose may not be clear, and, since this is a quick start, find the right balance between showing the user how to do _everything_, and showing them how to get up and running quickly.

File: 02_services_and_servicemanager.md [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+```
+
+As you can see we define two functions. The first being `findAllAlbums()` that is supposed to return all Albums and the
+second one being `findAlbum($id)` that is supposed to return the Album matching the given identifier `$id`. What's new
+in here is the fact that we actually define a return value that's not existing yet. We make the assumption that the
+return value all in all are of type `Album\Model\Album`. We will define this class at a later point and for now we can
+just create the `AlbumService` first.
+
+Create the class `AlbumService` at `/module/Album/src/Album/Service/AlbumService.php`, be sure to implement the
+`AlbumServiceInterface` and it's required functions. You then should have a class that looks like the following:
```
Comment (tfountain) [inline 02_services_and_servicemanager.md]: Let's mention here that we're just putting in placeholders for the function bodies for now, otherwise the code sample below looks unfinished.

> Create the class `AlbumService` at `/module/Album/src/Album/Service/AlbumService.php`, be sure to implement the `AlbumServiceInterface` and it's required functions (we'll fill in the functions later). You then should have a class that looks like the following:

File: 02_services_and_servicemanager.md [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+    public function __construct(AlbumServiceInterface $albumService)
+    {
+        $this->albumService = $albumService;
+    }
+}
+```
+
+As you can see our `__construct()` function now has a required argument. We will not be able to call this class anymore
+without passing it an instance of a class that matches our defined `AlbumServiceInterface`. If you were to go back to
+your browser and reload your project with the url `domain.loc/album`, you'd see the following error message:
```
Comment (tfountain) [inline 02_services_and_servicemanager.md:383]: "We will not be able to call this class" is confusing here, as "we" don't call the controller class directly. Perhaps:

> As you can see, our `__construct()` function now has the album service as a required argument. The application can only create an instance of this controller if it supplies an instance of `AlbumServiceInterface` as an argument. If you were to go back to your browser and reload your project with the URL `domain.loc/album`, you'd see the following error message:

**Comment by manuakasam** [comment]
I actually agree and this has become more clear to me when writing the third and fourth chapter. The interface makes more sense to define the datasource within the Service, not the Service itself. Will be refactored at a later point tho.

**Comment by manuakasam** [comment]
But then again, you don't want to have required implementations within any constructor :S I don't know, guess I'll wait for other feedback regarding that.

File: 02_services_and_servicemanager.md [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+```text
+( ! ) Catchable fatal error: Argument 1 passed to Album\Controller\ListController::__construct()
+      must be an instance of Album\Service\AlbumServiceInterface, none given,
+      called in {libraryPath}\Zend\ServiceManager\AbstractPluginManager.php on line {lineNumber}
+      and defined in \module\Album\src\Album\Controller\ListController.php on line 15
+```
+
+And this error message is expected. It tells you exactly that our `ListController` expects to be passed an implementation
+of the `AlbumServiceInterface`. So how do we make sure that our `ListController` will receive such an implementation?
+The solution to this lies within our definition of our controller inside the `module.config.php`. Let's review:
```
Comment (tfountain) [inline 02_services_and_servicemanager.md]: Some of this just repeats what was said above, and the writing style of posing a question that is then answered doesn't quite fit with the rest of the text. How about:

> To solve this, we need to tell the application how to create instances of the `Album\Controller\ListController`. If you remember back to when we created the controller, we added an entry to the `invokables` array in the module config:

File: 02_services_and_servicemanager.md [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+
+As you can see our controller name `Album\Controller\List` is defined under the key `invokables`. Whenever you place
+a class under the key `invokables` you let the [`Zend\Mvc\Controller\ControllerManager`](https://github.com/zendframework/zf2/blob/master/library/Zend/Mvc/Controller/ControllerManager.php)
+know that it should call the class without wiring in any hard written dependencies. In other words: our `ListController`
+can not be an `invokable` because it does have a hard written dependency called `AlbumServiceInterface`.
+
+The [`Zend\Mvc\Controller\ControllerManager`](https://github.com/zendframework/zf2/blob/master/library/Zend/Mvc/Controller/ControllerManager.php)
+does listen to another key, too, called `factories`. Factories are classes that are called from the Managers of the
+Zend Framework Application. Every factory-class is supposed to have a return value. Usually this return value is a class.
+Let's modify our configuration like this:
```
Comment (tfountain) [inline 02_services_and_servicemanager.md]: I think both of these paragraphs are a bit wordy. All we need to do here is explain what an invokable is (it should then be obvious why that will no longer work), and then explain factories. I'd suggest replacing these two paragraphs with:

> An `invokable` is a class that can be constructed without any arguments. Since our `Album\Controller\ListController` now has a required argument, we need to change this. The controller manager also support using _factories_. A factory is a class that creates instances of another class. We'll create one for our list controller. Let's modify our configuration like this:

Perhaps get someone to check my explanations there, and we might need to bring up the controller manager somewhere earlier on.

Also, personally I would just do this as a closure in the module's `getControllerConfig()` - do we have any explanations on when to use factory classes and when to use closures?

Comment (manuakasam) [inline 02_services_and_servicemanager.md]: There's no difference really. It's just an alternate approach. Personally, I use Factory-Classes whenever they are tied to creating a class that I own.

In the next chapter I think you'll see me using getServiceConfig() to create the TableGateway instance. This is due to me not owning TableGateway, so creating a 'Album\Db\TableGateway\TableGatewayFactory' would be kinda overkill :)

But that's how I justify the difference :P Ultimately, whenever you have FEW dependencies a closure is totally fine. Only if you have lots of stuff going on inside the factory then outsourcing it should be favored.

File: 02_services_and_servicemanager.md [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+Refreshing the browser won't change anything yet because we obviously need to modify our view-file to be able to display
+the data we want to.
+
+
+Accessing View Variables
+========================
+
+When pushing variables to the view they are accessible through two ways. Either directly like `$this->albums` or
+implicitly like `$albums`. Both are the same however calling `$albums` implicitly will result in a little round-trip
+through the `__call()` function which theoretically is a little slower. You won't ever notice this though.
```
Comment (tfountain) [inline 02_services_and_servicemanager.md:603]: Do we even need to mention this, as our examples use `$this->albums`? I didn't even know you could access var another way. Not sure if we need to introduce a concept that is both slower and not going to be used in this tutorial.

File: 02_services_and_servicemanager.md (hunk 1 above)
Comment (manuakasam) [inline 02_services_and_servicemanager.md]: agreed :thumbsup:

File: 02_services_and_servicemanager.md (hunk 3 above)
Comment (manuakasam) [inline 02_services_and_servicemanager.md]: agreed, sounds much better :thumbsup:

File: 02_services_and_servicemanager.md (hunk 4 above)
Comment (manuakasam) [inline 02_services_and_servicemanager.md]: Implemented your brush up of the paragraphs but stuck to the factory class still.

File: 02_services_and_servicemanager.md (hunk 5 above)
Comment (manuakasam) [inline 02_services_and_servicemanager.md:603]: I'd love feedback from others on this. Personally I think it is noteworthy, but if it appears confusing to others, too, I'll remove it.

Comment (gromson) [inline 02_services_and_servicemanager.md:603]: As for me this is a useful note. Because I knew that I can access to variable implicitly and directly, but I must have missed what the difference between these two ways during the reading an official documentation.

File: 02_services_and_servicemanager.md [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+     *
+     * @return string
+     */
+    public function getArtist();
+}
+```
+
+Notice that we only created getter-functions here. This is because right now we don't bother how the data gets inside
+the Album-class. All we care for is that we're able to access the properties through these getter-functions.
+
```
Comment (haclong) [inline 02_services_and_servicemanager.md:131]: ok, always write an interface before coding a concrete implementation of that said interface... but it is not clear WHY we have to write interface. From a OOP noob point of view, it sounds like writing the same thing twice.
Plus, we are here in a one-to-one example... 
i mean, writing an interface for several different concrete implementation does make sense (somehow, but then again, a total OOP noob need time to digest that info) but it is not clear in a one-to-one implementation...

File: 02_services_and_servicemanager.md (hunk 5 above)
Comment (haclong) [inline 02_services_and_servicemanager.md:603]: i don't feel it really confusing and it is good to know
i'll keep it if you'd asked me
